src/replace.cu

```diff
@@ -0,0 +1,174 @@
+          thrust::device, to_replace_begin, to_replace_end, data[i]);
+
+        if (found_ptr != to_replace_end) {
+            typename thrust::iterator_traits<
+              const thrust::device_ptr<const T>>::difference_type
+              value_found_index = thrust::distance(to_replace_begin, found_ptr);
+
+            data[i] = values[value_found_index];
+        }
+    }
+}
+
+/// /brief Call replace kernel according to primitive type T
+/// \param[in/out] data with elements to be replaced
+/// \param[in] data_size number of elements in data
+/// \param[in] to_replace contains values that will be replaced
+/// \param[in] values contains the replacement values
+/// \param[in] replacement_ptrdiff to get the end pointer of `to_replace` array
+template <class T>
+static inline gdf_error
+Replace(T *const             data,
+        const std::size_t    data_size,
+        const T *const       to_replace,
+        const T *const       values,
+        const std::ptrdiff_t replacement_ptrdiff) {
+    const std::size_t blocks = std::ceil(data_size / 256.);
+
+    const thrust::device_ptr<const T> to_replace_begin(to_replace);
+    const thrust::device_ptr<const T> to_replace_end(to_replace_begin
+                                                     + replacement_ptrdiff);
+
+    replace_kernel<T><<<blocks, 256>>>(  // TODO: calc blocks and threads
+      data,
+      data_size,
+      values,
+      to_replace_begin,
+      to_replace_end);
+
+    return GDF_SUCCESS;
+}
+
+/// \brief Check if two gdf_columns have the same size
+/// \param[in] to_replace is a gdf_column
+/// \param[in] values is a gdf_column
+static inline bool
+NotEqualReplacementSize(const gdf_column *to_replace,
+                        const gdf_column *values) {
+    return to_replace->size != values->size;
+}
+
+/// \brief Check if the three gdf columns have the same dtype
+/// \param[in] column is as gdf_column
+/// \param[in] to_replace is a gdf_column
+/// \param[in] values is a gdf_column
+static inline bool
+NotSameDType(const gdf_column *column,
+             const gdf_column *to_replace,
+             const gdf_column *values) {
+    return column->dtype != to_replace->dtype
+           || to_replace->dtype != values->dtype;
+}
+
+}  // namespace
+
+/// \brief Replace `to_replace` data of `column` with `values`
+/// \param[in/out] column data
+/// \param[in] to_replace contains values of column that will be replaced
+/// \param[in] values contains the replacement values
+///
+/// Note that `to_replace` and `values` are related by the index
+gdf_error
+gdf_replace(gdf_column *      column,
+            const gdf_column *to_replace,
+            const gdf_column *values) {
+    if (NotEqualReplacementSize(to_replace, values)) {
+        return GDF_COLUMN_SIZE_MISMATCH;
+    }
+
+    if (NotSameDType(column, to_replace, values)) { return GDF_CUDA_ERROR; }
+
+    switch (column->dtype) {
+#define WHEN(DTYPE)                                                            \
```

Comment:
I don't really like calling this macro "WHEN" -- it's not descriptive. Something like `REPLACE_CASE` would be clearer. Then you would have 

```
switch (column->type) {
    REPLACE_CASE(INT8);
    REPLACE_CASE(INT16);
    // etc.
}
```